tests/test_async_retry.py

```diff
@@ -0,0 +1,134 @@
+            };
+        };
+    '''
+
+    TEARDOWN = '''
+        DROP TYPE test::Counter;
+    '''
+
+    async def test_async_retry_01(self):
+        async for tx in self.con.retry():
+            async with tx:
+                await tx.execute('''
+                    INSERT test::Counter {
+                        name := 'counter1'
+                    };
+                ''')
+
+    async def test_async_retry_conflict(self):
+        con2 = await self.connect(database=self.get_database_name())
+        self.addCleanup(con2.aclose)
+
+        barrier = Barrier(2)
+        iterations = 0
+
+        async def transaction1(con):
+            async for tx in con.retry():
+                nonlocal iterations
+                iterations += 1
+                async with tx:
+                    # This magic query makes the test more reliable for some
+                    # reason. I guess this is because starting a transaction
+                    # in EdgeDB (and/or Postgres) is accomplished somewhat
+                    # lazily, i.e. only start transaction on the first query
+                    # rather than on the `START TRANSACTION`.
+                    await tx.query("SELECT 1")
+
+                    # Start both transactions at the same initial data.
+                    # One should succeed other should fail and retry.
+                    # On next attempt, the latter should succeed
+                    await barrier.ready()
+
+                    res = await tx.query_one('''
+                        SELECT (
+                            INSERT test::Counter {
+                                name := 'counter2',
+                                value := 1,
+                            } UNLESS CONFLICT ON .name
+                            ELSE (
+                                UPDATE test::Counter
+                                SET { value := .value + 1 }
+                            )
+                        ).value
+                    ''')
+            return res
+
+        results = await asyncio.wait_for(asyncio.gather(
+            transaction1(self.con),
+            transaction1(con2),
+            return_exceptions=True,
+        ), 10)
+        for e in results:
+            if isinstance(e, BaseException):
+                log.exception("Coroutine exception", exc_info=e)
+
+        self.assertEqual(set(results), {1, 2})
+        self.assertEqual(iterations, 3)
+
+    async def test_async_transaction_interface_errors(self):
+        with self.assertRaisesRegex(edgedb.InterfaceError,
+                                    r'.*the transaction is already started'):
+            async for tx in self.con.retry():
+                async with tx:
+                    await tx.start()
+
+        with self.assertRaisesRegex(edgedb.InterfaceError,
+                                    r'.*Use `async with transaction:`'):
+            async for tx in self.con.retry():
+                await tx.start()
```

Comment:
Can you add a test that checks what happens if an unrelated exception is propagated out of the body of with?

e.g.:

```python
async for tx in self.con.retry():
  async with tx:
    1/0
```

Reply:
I'm not sure what should be tested, though. Exception propagation? The successful rollback? No retries?

Reply:
> Exception propagation?

Yes. A very basic test that the iteration aborts and that the exception is propagated unmodified.

> The successful rollback? 

Yes, that too.

Reply:
Just wanted to add that tests like this, while trivial, helped me a bunch of times in the past. Just adding `return True` somewhere in `__exit__` can break everything and it might not be easy to spot that in review. So let's have some very basic tests to ensure the correct behavior.

Reply:
Done